`install.sh` script intermittently fails when trying to fetch releases from GitHub API

Opening this issue on behalf of a customer.

## Issue
When a GitHub workflow tries to download Coder, the workflow sometimes fails. The workflow is able to download the `install.sh` script, but [line 106 of the script](https://github.com/coder/coder/blob/426b348560cb9b4b9e2bc5e8c971e0761adebf9f/install.sh#L106) breaks because it tries to download an invalid release.

## Error logs
All IP addresses have been scrubbed from the output.

```txt
2024-12-11T00:10:01.7299771Z --2024-12-11 00:10:01-- https://coder.com/install.sh
2024-12-11T00:10:01.7465128Z Resolving coder.com (coder.com)... <series of IP addresses>, ...
2024-12-11T00:10:01.7556746Z Connecting to coder.com (coder.com)|<IP address>... connected.
2024-12-11T00:10:01.8635575Z HTTP request sent, awaiting response... 302 Found
2024-12-11T00:10:01.8636503Z Location: https://raw.githubusercontent.com/coder/coder/main/install.sh [following]
2024-12-11T00:10:01.8637504Z --2024-12-11 00:10:01-- https://raw.githubusercontent.com/coder/coder/main/install.sh
2024-12-11T00:10:01.8702735Z Resolving raw.githubusercontent.com (raw.githubusercontent.com)... <IP addresses>
2024-12-11T00:10:01.8808504Z Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|<IP address>... connected.
2024-12-11T00:10:01.9163343Z HTTP request sent, awaiting response... 200 OK
2024-12-11T00:10:01.9164016Z Length: 20724 (20K) [text/plain]
2024-12-11T00:10:01.9164666Z Saving to: 'STDOUT'
2024-12-11T00:10:01.9164949Z 
2024-12-11T00:10:01.9172588Z 0K .......... .......... 100% 23.7M=0.001s
2024-12-11T00:10:01.9173029Z 
2024-12-11T00:10:01.9173369Z 2024-12-11 00:10:01 (23.7 MB/s) - written to stdout [20724/20724]
2024-12-11T00:10:01.9173795Z 
2024-12-11T00:10:02.5000981Z curl: (22) The requested URL returned error: 403
2024-12-11T00:10:02.5036592Z Ubuntu 22.04.5 LTS
2024-12-11T00:10:02.5053521Z Installing v of the amd64 deb package from GitHub.
2024-12-11T00:10:02.5064357Z 
2024-12-11T00:10:02.5077135Z + mkdir -p ~/.cache/coder
2024-12-11T00:10:02.5104499Z + curl -#fL -o ~/.cache/coder/coder__amd64.deb.incomplete -C - https://github.com/coder/coder/releases/download/v/coder%5F%5Flinux%5Famd64.deb
2024-12-11T00:10:02.7228897Z #=#=# 
2024-12-11T00:10:02.7229418Z curl: (22) The requested URL returned error: 404
2024-12-11T00:10:02.7229715Z 
2024-12-11T00:10:02.7265228Z ##[error]Process completed with exit code 22.
```

--- a/install.sh
+++ b/install.sh
@@ -92,8 +92,19 @@ EOF
 }
 
 echo_latest_stable_version() {
+	url="https://github.com/coder/coder/releases/latest"
 	# https://gist.github.com/lukechilds/a83e1d7127b78fef38c2914c4ececc3c#gistcomment-2758860
-	version="$(curl -fsSLI -o /dev/null -w "%{url_effective}" https://github.com/coder/coder/releases/latest)"
+	response=$(curl -sSLI -o /dev/null -w "\n%{http_code} %{url_effective}" ${url})
+	status_code=$(echo "$response" | tail -n1 | cut -d' ' -f1)
+	version=$(echo "$response" | tail -n1 | cut -d' ' -f2-)
+	body=$(echo "$response" | sed '$d')
+
+	if [ "$status_code" != "200" ]; then
+		echoerr "GitHub API returned status code: ${status_code}"
+		echoerr "URL: ${url}"
+		exit 1
+	fi
+
 	version="${version#https://github.com/coder/coder/releases/tag/v}"
 	echo "${version}"
 }
@@ -103,7 +114,19 @@ echo_latest_mainline_version() {
 	# and take the first result. Note that we're sorting by space-
 	# separated numbers and without utilizing the sort -V flag for the
 	# best compatibility.
-	curl -fsSL https://api.github.com/repos/coder/coder/releases |
+	url="https://api.github.com/repos/coder/coder/releases"
+	response=$(curl -sSL -w "\n%{http_code}" ${url})
+	status_code=$(echo "$response" | tail -n1)
+	body=$(echo "$response" | sed '$d')
+
+	if [ "$status_code" != "200" ]; then
+		echoerr "GitHub API returned status code: ${status_code}"
+		echoerr "URL: ${url}"
+		echoerr "Response body: ${body}"
+		exit 1
+	fi
+
+	echo "$body" |
 		awk -F'"' '/"tag_name"/ {print $4}' |
 		tr -d v |
 		tr . ' ' |
@@ -405,8 +428,10 @@ main() {
 	STABLE_VERSION=$(echo_latest_stable_version)
 	if [ "${MAINLINE}" = 1 ]; then
 		VERSION=$(echo_latest_mainline_version)
+		echoh "Resolved mainline version: v${VERSION}"
 	elif [ "${STABLE}" = 1 ]; then
 		VERSION=${STABLE_VERSION}
+		echoh "Resolved stable version: v${VERSION}"
 	fi
 
 	distro_name

